compiler/rustc_typeck/src/check/mod.rs

```diff
@@ -951,3 +952,32 @@ fn has_expected_num_generic_args<'tcx>(
         generics.count() == expected + if generics.has_self { 1 } else { 0 }
     })
 }
+
+/// Suggests calling the constructor of a tuple struct or enum variant
+///
+/// * `snippet` - The snippet of code that references the constructor
+/// * `span` - The span of the snippet
+/// * `params` - The number of parameters the constructor accepts
+/// * `err` - A mutable diagnostic builder to add the suggestion to
+fn suggest_call_constructor(
+    snippet: impl std::fmt::Display,
+    span: Span,
+    params: usize,
+    err: &mut DiagnosticBuilder<'_>,
+) {
+    // Note: tuple-structs don't have named fields, so just use placeholders
+    let args = vec!["_"; params].join(", ");
+    let applicable = if params > 0 {
+        Applicability::HasPlaceholders
+    } else {
+        // When n = 0, it's an empty-tuple struct/enum variant
+        // so we trivially know how to construct it
+        Applicability::MachineApplicable
+    };
+    err.help(&format!("`{}` is the constructor of a struct or enum variant", snippet));
```

Comment:
It seems like we could get the `Res` to be explicit on whether the expression is a `tuple-struct` or `tuple variant` constructor.